### PR TITLE
a number of fixes / tweaks to the diagnostics system

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

- Fixes a regression in the completion system, caused by the old refactor from `JSON::Value.parse()` returning an error rather than a `bool`
- Fixes a similar regression in the refactor here: https://github.com/rstudio/rstudio/commit/158deb46fd2ba9891647f817548ee6efac952535#diff-97d5752c526b19ac362ddcf254af529722e2983120482f44fed849d5cb9061c7R363
- Improves handling of `glue()` strings
- Adds a number of other special cases for how certain R functions handle missing arguments

## Diagnostics with glue

Screen Shot 2022-06-18 at 9 01 55 PM

### Approach

Mostly straightforward tweaks throughout the diagnostics system.

### Automated Tests

A test for `glue()` is included.

### QA Notes

> Add additional information for QA on how to validate the change, paying special attention to the level of risk, adjacent areas that could be affected by the change, and any important contextual information not present in the linked issues. 

### Checklist

- [ ] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [ ] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [ ] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [ ] This PR passes all local unit tests

### CODE REVIEW

**Comment by kevinushey** [comment]
Sorry, this PR is smaller than it looks -- a lot of the changes are just re-organizations of code.

File: src/cpp/session/modules/SessionRParser.cpp [hunk 1]
```diff
@@ -483,729 +493,393 @@ std::wstring& wideComplement(const std::wstring& bracket)
...
-   
-   Error error = r::sexp::extractFunctionInfo(
-            functionSEXP,
-            &info,
-            true,
-            true);
-   
-   return info;
-   
-}
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRParser.cpp:1168]: End of code re-organization.

File: src/cpp/session/modules/SessionRParser.cpp [hunk 2]
```diff
@@ -1274,220 +948,628 @@ class MatchedCall
...
+      }
+      
+      if (cursor.contentEquals(L"need"))
+         pCall->functionInfo().infoForFormal("label").setMissingnessHandled(true);
+      
+      if (cursor.contentEquals(L"quo"))
+         pCall->functionInfo().infoForFormal("expr").setMissingnessHandled(true);
+   }
+   
+   static void applyCustomWarnings(const MatchedCall& call,
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRParser.cpp:1124]: Start of code re-organization.

**Review by jgutman — APPROVED** [review]
LGTM although I'm not super familiar with the issue being addressed; test cases look good!

**Pull request merged**